server/src/main/java/org/opensearch/gateway/remote/routingtable/IndexRoutingTableInputStream.java

```diff
@@ -0,0 +1,174 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+
+package org.opensearch.gateway.remote.routingtable;
+
+import org.opensearch.Version;
+import org.opensearch.cluster.routing.IndexRoutingTable;
+import org.opensearch.cluster.routing.IndexShardRoutingTable;
+import org.opensearch.common.io.stream.BufferedChecksumStreamOutput;
+import org.opensearch.common.io.stream.BytesStreamOutput;
+import org.opensearch.core.common.bytes.BytesReference;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.util.Iterator;
+
+public class IndexRoutingTableInputStream extends InputStream {
+
+    /**
+     * The buffer where data is stored.
+     */
+    protected byte[] buf;
+
+    /**
+     * The number of valid bytes in the buffer.
+     */
+    protected int count;
+
+    /**
+     * The buffer left over from the last fill
+     */
+    protected byte[] leftOverBuf;
+
+    /**
+     * The mark position
+     */
+    protected int markPos = -1;
+
+    /**
+     * The read limit
+     */
+    protected int markLimit;
+
+    /**
+     * The position
+     */
+    protected int pos;
+
+    private static final int BUFFER_SIZE = 8192;
+
+    private final IndexRoutingTableHeader indexRoutingTableHeader;
+    private final Iterator<IndexShardRoutingTable> shardIter;
+    private final BytesStreamOutput bytesStreamOutput;
+    private final BufferedChecksumStreamOutput out;
+
+    public IndexRoutingTableInputStream(IndexRoutingTable indexRoutingTable, long version, Version nodeVersion) throws IOException {
+        this(indexRoutingTable, version, nodeVersion, BUFFER_SIZE);
+    }
+
+    public IndexRoutingTableInputStream(IndexRoutingTable indexRoutingTable, long version, Version nodeVersion, int size)
+        throws IOException {
+        this.buf = new byte[size];
+        this.shardIter = indexRoutingTable.iterator();
+        this.indexRoutingTableHeader = new IndexRoutingTableHeader(version, indexRoutingTable.getIndex().getName(), nodeVersion);
+        this.bytesStreamOutput = new BytesStreamOutput();
+        this.out = new BufferedChecksumStreamOutput(bytesStreamOutput);
```

Comment:
Why do we require them to be properties of the class. All that we should care about is the byte array that acts as a buffer to get reused across operations

Reply:
We need the checksum to be added for complete file, for this we need to ensure we reuse the same `BufferedChecksumStreamOutput` throughout the write operations.

Reply:
Exactly so `BufferedChecksumStreamOutput` should be used only within the context of a method or specific operation as a temp storage rather than exposing it as a class property

Reply:
The issue with keeping `BufferedChecksumStreamOutput` as a temp storage is we will then need to retain and update checksum independently. 
The usage of this class is something like:
`inputStream = new IndexRoutingTableInputStream()` -> which calls `initialFill` where we put some content in buffer.
Post this `inputStream.read()` will be called in a loop till we have content in buffer. 
In both these methods we need to retain the checksum and add to it with the next content read. `BufferedChecksumStreamOutput` does not currently expose any methods to add checksum to an initial value. 

The way I see we have 2 options here:
1. Keep using `BufferedChecksumStreamOutput` as class property which takes care of updating checksum across reads.
2. Retain checksum from one read and update `BufferedChecksumStreamOutput` to be able to use an initial value in the next read.

Please suggest if there is a better way of handling checksum or if I am missing something here.